

<!DOCTYPE html>

<html lang="en-US">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=Edge">

  

  <link rel="shortcut icon" href="/GLIMPSE/favicon.ico" type="image/x-icon">


  <link rel="stylesheet" href="/GLIMPSE/assets/css/just-the-docs-default.css">

  
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-2TSEMF47BK"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-2TSEMF47BK', { 'anonymize_ip': true });
    </script>

  

  
    <script type="text/javascript" src="/GLIMPSE/assets/js/vendor/lunr.min.js"></script>
  

  
    <script src="https://cdn.jsdelivr.net/npm/mermaid@9.1.6/dist/mermaid.min.js"></script>
  

  <script type="text/javascript" src="/GLIMPSE/assets/js/just-the-docs.js"></script>

  <meta name="viewport" content="width=device-width, initial-scale=1">

  <!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Glossary | GLIMPSE2</title>
<meta name="generator" content="Jekyll v4.2.2" />
<meta property="og:title" content="Glossary" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="A tool for low-coverage whole genome sequencing imputation" />
<meta property="og:description" content="A tool for low-coverage whole genome sequencing imputation" />
<link rel="canonical" href="http://localhost:4000/GLIMPSE/docs/glossary/" />
<meta property="og:url" content="http://localhost:4000/GLIMPSE/docs/glossary/" />
<meta property="og:site_name" content="GLIMPSE2" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Glossary" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebPage","description":"A tool for low-coverage whole genome sequencing imputation","headline":"Glossary","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"http://localhost:4000/GLIMPSE/assets/images/branding/glimpse_logo.svg"}},"url":"http://localhost:4000/GLIMPSE/docs/glossary/"}</script>
<!-- End Jekyll SEO tag -->


  

</head>

<body>
  <a class="skip-to-main" href="#main-content">Skip to main content</a>
  <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
    <symbol id="svg-link" viewBox="0 0 24 24">
      <title>Link</title>
      <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link">
        <path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path>
      </svg>
    </symbol>
    <symbol id="svg-search" viewBox="0 0 24 24">
      <title>Search</title>
      <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search">
        <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line>
      </svg>
    </symbol>
    <symbol id="svg-menu" viewBox="0 0 24 24">
      <title>Menu</title>
      <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu">
        <line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line>
      </svg>
    </symbol>
    <symbol id="svg-arrow-right" viewBox="0 0 24 24">
      <title>Expand</title>
      <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right">
        <polyline points="9 18 15 12 9 6"></polyline>
      </svg>
    </symbol>
    <symbol id="svg-doc" viewBox="0 0 24 24">
      <title>Document</title>
      <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file">
        <path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline>
      </svg>
    </symbol>
    <!-- Feather. MIT License: https://github.com/feathericons/feather/blob/master/LICENSE -->
<symbol id="svg-external-link" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-external-link">
  <title id="svg-external-link-title">(external link)</title>
  <path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path><polyline points="15 3 21 3 21 9"></polyline><line x1="10" y1="14" x2="21" y2="3"></line>
</symbol>

  </svg>

  <div class="side-bar">
    <div class="site-header">
      <a href="/GLIMPSE/" class="site-title lh-tight">
  <div class="site-logo"></div>

</a>
      <a href="#" id="menu-button" class="site-button">
        <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-menu"></use></svg>
      </a>
    </div>
    <nav role="navigation" aria-label="Main" id="site-nav" class="site-nav">
      
      
        <ul class="nav-list"><li class="nav-list-item"><a href="/GLIMPSE/" class="nav-list-link">Home</a></li><li class="nav-list-item"><a href="#" class="nav-list-expander" aria-label="toggle links in Installation category">
            <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg>
          </a><a href="/GLIMPSE/docs/installation" class="nav-list-link">Installation</a><ul class="nav-list "><li class="nav-list-item "><a href="#" class="nav-list-expander" aria-label="toggle links in Build from source category">
                  <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg>
                </a><a href="/GLIMPSE/docs/installation/build_from_source" class="nav-list-link">Build from source</a><ul class="nav-list"><li class="nav-list-item ">
                    <a href="/GLIMPSE/docs/installation/build_from_source/system_requirements" class="nav-list-link">System requirements</a>
                  </li><li class="nav-list-item ">
                    <a href="/GLIMPSE/docs/installation/build_from_source/required_libraries" class="nav-list-link">Required libraries</a>
                  </li><li class="nav-list-item ">
                    <a href="/GLIMPSE/docs/installation/build_from_source/compile_glimpse2" class="nav-list-link">Compile GLIMPSE2</a>
                  </li></ul></li><li class="nav-list-item "><a href="/GLIMPSE/docs/installation/static_binaries" class="nav-list-link">Static binaries</a></li><li class="nav-list-item "><a href="/GLIMPSE/docs/installation/docker" class="nav-list-link">Docker</a></li></ul></li><li class="nav-list-item"><a href="#" class="nav-list-expander" aria-label="toggle links in Documentation category">
            <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg>
          </a><a href="/GLIMPSE/docs/documentation" class="nav-list-link">Documentation</a><ul class="nav-list "><li class="nav-list-item "><a href="/GLIMPSE/docs/documentation/chunk/" class="nav-list-link">chunk</a></li><li class="nav-list-item "><a href="/GLIMPSE/docs/documentation/phase/" class="nav-list-link">phase</a></li><li class="nav-list-item "><a href="/GLIMPSE/docs/documentation/split_reference/" class="nav-list-link">split_reference</a></li><li class="nav-list-item "><a href="/GLIMPSE/docs/documentation/ligate/" class="nav-list-link">ligate</a></li><li class="nav-list-item "><a href="/GLIMPSE/docs/documentation/concordance/" class="nav-list-link">concordance</a></li></ul></li><li class="nav-list-item"><a href="#" class="nav-list-expander" aria-label="toggle links in Tutorials category">
            <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg>
          </a><a href="/GLIMPSE/docs/tutorials" class="nav-list-link">Tutorials</a><ul class="nav-list "><li class="nav-list-item "><a href="/GLIMPSE/docs/tutorials/ukb_wgs/" class="nav-list-link">UK Biobank WGS data</a></li><li class="nav-list-item "><a href="/GLIMPSE/docs/tutorials/ukb_phasing/" class="nav-list-link">Phasing</a></li></ul></li><li class="nav-list-item"><a href="/GLIMPSE/docs/quickstart/" class="nav-list-link">Expert quick start</a></li><li class="nav-list-item active"><a href="/GLIMPSE/docs/glossary/" class="nav-list-link active">Glossary</a></li><li class="nav-list-item"><a href="/GLIMPSE/CHANGELOG/" class="nav-list-link">CHANGELOG</a></li><li class="nav-list-item external">
      <a href="https://github.com/odelaneau/GLIMPSE" class="nav-list-link external">
        GLIMPSE2 on GitHub
        <svg viewBox="0 0 24 24" aria-labelledby="svg-external-link-title"><use xlink:href="#svg-external-link"></use></svg>
      </a>
    </li><li class="nav-list-item external">
      <a href="https://odelaneau.github.io/GLIMPSE/glimpse1" class="nav-list-link external">
        GLIMPSE1 website
        <svg viewBox="0 0 24 24" aria-labelledby="svg-external-link-title"><use xlink:href="#svg-external-link"></use></svg>
      </a>
    </li></ul>
      
      
    </nav>

    
    
       

    
  </div>
  <div class="main" id="top">
    <div id="main-header" class="main-header">
      

        

        <div class="search">
          <div class="search-input-wrap">
            <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search GLIMPSE2" aria-label="Search GLIMPSE2" autocomplete="off">
            <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label>
          </div>
          <div id="search-results" class="search-results"></div>
        </div>
      
      
      
        <nav aria-label="Auxiliary" class="aux-nav">
          <ul class="aux-nav-list">
            
              <li class="aux-nav-list-item">
                <a href="//github.com/odelaneau/GLIMPSE" class="site-button"
                  
                  target="_blank" rel="noopener noreferrer"
                  
                >
                  GLIMPSE2 on GitHub
                </a>
              </li>
            
          </ul>
        </nav>
      
    </div>
    <div id="main-content-wrap" class="main-content-wrap">
      
        
      
      <div id="main-content" class="main-content" role="main">
        
          <h1 id="glossary-of-terms">
        
        
          <a href="#glossary-of-terms" class="anchor-heading" aria-labelledby="glossary-of-terms"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Glossary of Terms
        
        
      </h1>
    

<p>Below are some important concepts, software, standards, and other things you might encounter in these Docs and working with SHAPEIT5.</p>

<ul>
  <li><a href="#glimpse">GLIMPSE</a></li>
  <li><a href="#genotype-imputation">Genotype imputation</a></li>
  <li><a href="#haplotype-phasing">Haplotype phasing</a></li>
  <li><a href="#haplotype-scaffold">Haplotype scaffold</a></li>
  <li><a href="#impute">IMPUTE</a></li>
  <li><a href="#low-coverage-whole-genome-sequencing">Low-coverage whole genome sequencing</a></li>
  <li><a href="#reference-panel">Reference panel</a></li>
  <li><a href="#shapeit">SHAPEIT</a></li>
  <li><a href="#snp-array">SNP array</a></li>
  <li><a href="#singleton">Singleton</a></li>
  <li><a href="#switch-error-rate">Switch error rate</a></li>
  <li><a href="#uk-biobank-research-analysis-platform">UK Biobank Research Analysis Platform</a></li>
  <li><a href="#whole-exome-sequencing-wes">Whole exome sequencing (WES)</a></li>
  <li><a href="#whole-genome-sequencing-wgs">Whole genome sequencing (WGS)</a></li>
</ul><hr />
      <h2 id="glimpse">
        
        
          <a href="#glimpse" class="anchor-heading" aria-labelledby="glimpse"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> GLIMPSE
        
        
      </h2>
    

<p>GLIMPSE (Genotype Likelihoods IMputation and PhaSing mEthod) is a method for haplotype phasing and genotype imputation of low-coverage WGS data, developed by Simone Rubinacci and Olivier Delaneau. The latest version of the tool is GLIMPSE2, which offers the best accuracy and computational performance for large reference panels at rare variants.</p>

<p>Resources:</p>
<ul>
  <li>GLIMPSE2 <a href="https://odelaneau.github.io/GLIMPSE/" target="_blank" rel="noopener">[Website]</a> <a href="https://www.biorxiv.org/content/10.1101/2022.11.28.518213v1">[Paper]</a></li>
  <li>GLIMPSE1 <a href="https://odelaneau.github.io/GLIMPSE/glimpse1/" target="_blank" rel="noopener">[Website]</a> <a href="https://www.nature.com/articles/s41588-020-00756-0">[Paper]</a></li>
</ul><hr />
      <h2 id="genotype-imputation">
        
        
          <a href="#genotype-imputation" class="anchor-heading" aria-labelledby="genotype-imputation"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Genotype imputation
        
        
      </h2>
    

<p>Genotype imputation is a probabilistic inference of the genotypes, typically used for SNP array data or low-coverage WGS. Imputation uses a large reference panel of phased haplotypes to determine large shared identity-by-descent segments that can be used for the statistical inference. Two widely used tools for genotype imputation are <a href="https://www.dropbox.com/sh/mwnceyhir8yze2j/AADbzP6QuAFPrj0Z9_I1RSmla?dl=0" target="_blank" rel="noopener">IMPUTE5 for SNP array data</a>  and <a href="https://odelaneau.github.io/GLIMPSE/" target="_blank" rel="noopener">GLIMPSE tool for low-coverage imputation</a>.</p>

<p>Resources:</p>
<ul>
  <li>GLIMPSE for low-coverage imputation <a href="https://odelaneau.github.io/GLIMPSE/" target="_blank" rel="noopener">[Website]</a> <a href="https://www.nature.com/articles/s41588-020-00756-0">[Paper]</a></li>
  <li>IMPUTE5 for SNP array data <a href="https://www.dropbox.com/sh/mwnceyhir8yze2j/AADbzP6QuAFPrj0Z9_I1RSmla?dl=0" target="_blank" rel="noopener">[Website]</a> <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1009049">[Paper]</a></li>
</ul><hr />
      <h2 id="haplotype-phasing">
        
        
          <a href="#haplotype-phasing" class="anchor-heading" aria-labelledby="haplotype-phasing"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Haplotype phasing
        
        
      </h2>
    

<p>Haplotype phasing involves distinguishing the two parentally inherited copies of each chromosome into haplotypes. <a href="https://odelaneau.github.io/shapeit5/" target="_blank" rel="noopener">SHAPEIT5</a> is a tool for statistical population-based haplotype phasing which focuses specifically at extremely rare variants.</p>

<p>Resources:</p>
<ul>
  <li>SHAPEIT5 <a href="https://odelaneau.github.io/shapeit5/" target="_blank" rel="noopener">[Website]</a> <a href="https://www.biorxiv.org/content/10.1101/2022.10.19.512867v1" target="_blank" rel="noopener">[Paper]</a></li>
</ul><hr />
      <h2 id="haplotype-scaffold">
        
        
          <a href="#haplotype-scaffold" class="anchor-heading" aria-labelledby="haplotype-scaffold"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Haplotype scaffold
        
        
      </h2>
    

<p>A haplotype scaffold is a set of highly confident haplotypes, typically on a subset of the data. SHAPEIT5 uses the haplotypes derived at common variants as haplotype scaffolds onto which heterozygous genotypes are phased one rare variant at a time.</p><hr />
      <h2 id="impute">
        
        
          <a href="#impute" class="anchor-heading" aria-labelledby="impute"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> IMPUTE
        
        
      </h2>
    

<p>IMPUTE is a method for genotype imputation of SNP array data, developed by Jonathan Marchini, Bryan Howie and Simone Rubinacci. The latest version of the tool is IMPUTE5, which offers efficient computational performances for large reference panels.</p>

<p>Resources:</p>

<ul>
  <li>IMPUTE5 <a href="https://jmarchini.org/software/" target="_blank" rel="noopener">[Website]</a> <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1009049">[Paper]</a></li>
  <li>IMPUTE4 <a href="https://jmarchini.org/software/" target="_blank" rel="noopener">[Website]</a> <a href="https://www.nature.com/articles/s41586-018-0579-z">[Paper]</a></li>
  <li>IMPUTE2 <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1000529" target="_blank" rel="noopener">[Paper 1]</a> <a href="https://www.nature.com/articles/ng.2354">[Paper 2]</a></li>
  <li>IMPUTE1 <a href="https://www.nature.com/articles/ng2088" target="_blank" rel="noopener">[Paper]</a></li>
</ul><hr />
      <h2 id="low-coverage-whole-genome-sequencing">
        
        
          <a href="#low-coverage-whole-genome-sequencing" class="anchor-heading" aria-labelledby="low-coverage-whole-genome-sequencing"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Low-coverage whole genome sequencing
        
        
      </h2>
    

<p>Low-coverage whole genome sequencing is whole genome sequencing data sequenced at low-depth usually data with a coverage between 0.1x-8x. Conversely to high-coverage WGS, genotype calling in this case is uncertain and very often this data needs to be imputed with specialised methods (e.g. <a href="https://odelaneau.github.io/GLIMPSE/" target="_blank" rel="noopener">GLIMPSE</a>) in order to be used and processed.</p>

<p>Resources:</p>
<ul>
  <li>GLIMPSE for low-coverage WGS imputation <a href="https://odelaneau.github.io/GLIMPSE/" target="_blank" rel="noopener">[Website]</a> <a href="https://www.biorxiv.org/content/10.1101/2022.10.19.512867v1" target="_blank" rel="noopener">[Paper]</a></li>
</ul><hr />
      <h2 id="reference-panel">
        
        
          <a href="#reference-panel" class="anchor-heading" aria-labelledby="reference-panel"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Reference panel
        
        
      </h2>
    

<p>A reference panel is a large set of deeply-sequenced haplotypes used for genotype imputation and haplotype phasing typically for cohorts genotyped using SNP array or low-coverage WGS.</p><hr />
      <h2 id="shapeit">
        
        
          <a href="#shapeit" class="anchor-heading" aria-labelledby="shapeit"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> SHAPEIT
        
        
      </h2>
    

<p>SHAPEIT (Segmented HAPlotype Estimation and Imputation Tools) is a commonly used method for haplotype phasing, mainly developed by Olivier Delaneau. The latest version of the tool is SHAPEIT5, which offers the best accuracy for large cohorts at rare variants.</p>

<p>Resources:</p>
<ul>
  <li>SHAPEIT5 <a href="https://odelaneau.github.io/SHAPEIT5/" target="_blank" rel="noopener">[Paper]</a></li>
  <li>SHAPEIT4 <a href="https://www.nature.com/articles/s41467-019-13225-y" target="_blank" rel="noopener">[Paper]</a></li>
  <li>SHAPEIT3 <a href="https://www.nature.com/articles/ng.3583" target="_blank" rel="noopener">[Paper]</a></li>
  <li>SHAPEIT2 <a href="https://www.nature.com/articles/nmeth.2307" target="_blank" rel="noopener">[Paper]</a></li>
  <li>SHAPEIT <a href="https://www.nature.com/articles/nmeth.1785" target="_blank" rel="noopener">[Paper]</a></li>
</ul><hr />
      <h2 id="snp-array">
        
        
          <a href="#snp-array" class="anchor-heading" aria-labelledby="snp-array"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> SNP array
        
        
      </h2>
    

<p>SNP array is a DNA microarray which is used to detect common SNPs within a population. In the context of GWAS, SNP array data is often imputed using a reference panel of haplotypes (e.g. using <a href="https://www.dropbox.com/sh/mwnceyhir8yze2j/AADbzP6QuAFPrj0Z9_I1RSmla?dl=0">IMPUTE5</a></p>

<p>Resources:</p>
<ul>
  <li>IMPUTE5 for SNP array data <a href="https://www.dropbox.com/sh/mwnceyhir8yze2j/AADbzP6QuAFPrj0Z9_I1RSmla?dl=0" target="_blank" rel="noopener">[Website]</a> <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1009049">[Paper]</a></li>
</ul><hr />
      <h2 id="singleton">
        
        
          <a href="#singleton" class="anchor-heading" aria-labelledby="singleton"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Singleton
        
        
      </h2>
    

<p>A singleton is a rare variant for which genetic variation is carried by a unique chromosome in the dataset (minor allele count of 1). As these variants are unique in the datasets, there is no information to phase the variant, therefore statistical population-based phasing has always reported a switch error rate of 50% at these sites. The typical way to phase these variants is to use family information. However, <a href="https://odelaneau.github.io/shapeit5/" target="_blank" rel="noopener">SHAPEIT5</a> is able to provide for the first time non-random phasing at these sites without the need of family information, by using a simple coalescent-inspired model.</p>

<p>Resources</p>
<ul>
  <li>SHAPEIT5 <a href="https://odelaneau.github.io/shapeit5/" target="_blank" rel="noopener">[Website]</a> <a href="https://www.biorxiv.org/content/10.1101/2022.10.19.512867v1" target="_blank" rel="noopener">[Paper]</a></li>
</ul><hr />
      <h2 id="switch-error-rate">
        
        
          <a href="#switch-error-rate" class="anchor-heading" aria-labelledby="switch-error-rate"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Switch error rate
        
        
      </h2>
    

<p>Switch error rate (SER) is a popular metric to measure the quality of phased data. SER is used when known maternal and paternal haplotypes are available and is defined as the number of switch errors divided by the number of opportunities for switch errors.</p><hr />
      <h2 id="uk-biobank-research-analysis-platform">
        
        
          <a href="#uk-biobank-research-analysis-platform" class="anchor-heading" aria-labelledby="uk-biobank-research-analysis-platform"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> UK Biobank Research Analysis Platform
        
        
      </h2>
    

<p>The UK Biobank Research Analysis Platform (RAP), enabled by DNAnexus and Amazon Web Services (AWS), is a cloud-based platform that enables researchers to work with the UK Biobank WGS and WES data.</p>

<p>Resources:</p>
<ul>
  <li><a href="https://www.ukbiobank.ac.uk/enable-your-research/research-analysis-platform" target="_blank" rel="noopener">UK Biobank website</a></li>
</ul><hr />
      <h2 id="whole-exome-sequencing-wes">
        
        
          <a href="#whole-exome-sequencing-wes" class="anchor-heading" aria-labelledby="whole-exome-sequencing-wes"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Whole exome sequencing (WES)
        
        
      </h2>
    

<p>Whole exome sequencing (WES) is a technique for sequencing all of the protein-coding regions of genes in a genome. In the UK Biobank dataset, we combined WES with SNP array data and performed phasing with SHAPEIT5.</p><hr />
      <h2 id="whole-genome-sequencing-wgs">
        
        
          <a href="#whole-genome-sequencing-wgs" class="anchor-heading" aria-labelledby="whole-genome-sequencing-wgs"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Whole genome sequencing (WGS)
        
        
      </h2>
    

<p>Whole-genome sequencing (WGS) is the process of determining the DNA sequence of an organism at a single time. WGS is intended at high coverage, from 30x of coverage or more, therefore the quality of genotype calling is high. <a href="https://odelaneau.github.io/shapeit5" target="_blank" rel="noopener">SHAPEIT5</a> is designed for large WGS cohorts (e.g. in the UK Biobank), to distinguish the two inherited copies of each chromosome into haplotypes (see haplotype phasing).</p>

        

        

        
        
          <hr>
          <footer>
            
              <p><a href="#top" id="back-to-top">Back to top</a></p>
            

            <p class="text-small text-grey-dk-100 mb-0">Copyright &copy; 2022-2023 Simone Rubinacci & Olivier Delaneau | All Rights Reserved | Distributed by an <a href="https://github.com/odelaneau/GLIMPSE/blob/master/LICENSE">MIT license.</a></p>

            
              <div class="d-flex mt-2">
                
                
                  <p class="text-small text-grey-dk-000 mb-0">
                    <a href="https://github.com/srubinacci/ideal-barnacle/tree/main/docs/docs/glossary.md" id="edit-this-page">Edit this page on GitHub</a>
                  </p>
                
              </div>
            
          </footer>
        

      </div>
    </div>

    
      

      <div class="search-overlay"></div>
    
  </div>
</body>

  <script>
    var config = {}
;
    mermaid.initialize(config);
    window.mermaid.init(undefined, document.querySelectorAll('.language-mermaid'));
  </script>

</html>

